### ISSUE TITLE
Duplicated logs uploaded from disk to cloud server

### ISSUE THREAD

**Issue by melix-sun** [rsyslog/rsyslog#5419]
### Expected behavior
All logs are sync to cloud server in order and no repeated
### Actual behavior
Some part of logs are sync repeatedly
### Steps to reproduce the behavior
1. use disk-assisted in-memory queue to sync logs to cloud server
2. logs are written into disk
3. logs in disk are sync to cloud server
4. restart vehicle
5. some part of logs are sync repeatedly
### Environment
- rsyslog version: 
- platform: 
- for configuration questions/issues, include rsyslog.conf and included config files

please refer to the attachment for the configuration (duplication details) and debug result (rsyslog-debug), and please kindly help to look at this issue.
[rsyslog-debug.zip](https://github.com/user-attachments/files/16358009/rsyslog-debug.zip)

 
[Duplication details.xlsx](https://github.com/user-attachments/files/16358008/Duplication.details.xlsx)

if you need more details to figure it out, please kindly let me know, thanks.

**Comment by davidelang** [comment]
when you use multiple threads, or UDP network communications, the order of 
messages is not always maintained.

When you use disk assisted queues, messages that spill to disk will be sent in 
parallel to messages in memory (and at a much slower rate) so you are guaranteed 
to lose order

with RELP communications, there are error conditions that can cause message 
duplication (when things fail, you can either assume they were delivered, or 
that they weren't delivered, so you either end up with message loss or message 
duplication, rsyslog has opted for message duplication)

I don't know what message you are using to sync the disk files to the remote 
server, so I can't comment on the message duplication that is happening with 
that method. It sounds as if you are using some method outside of rsyslog.

We would need to know a lot more about your config to comment much, but it 
sounds like you are getting misordering due to the expected behavior of rsyslog, 
and duplication due to the way that you are syncing the files on disk outside of 
rsyslog

David Lang

**Comment by melix-sun** [comment]
> when you use multiple threads, or UDP network communications, the order of messages is not always maintained. When you use disk assisted queues, messages that spill to disk will be sent in parallel to messages in memory (and at a much slower rate) so you are guaranteed to lose order with RELP communications, there are error conditions that can cause message duplication (when things fail, you can either assume they were delivered, or that they weren't delivered, so you either end up with message loss or message duplication, rsyslog has opted for message duplication) I don't know what message you are using to sync the disk files to the remote server, so I can't comment on the message duplication that is happening with that method. It sounds as if you are using some method outside of rsyslog. We would need to know a lot more about your config to comment much, but it sounds like you are getting misordering due to the expected behavior of rsyslog, and duplication due to the way that you are syncing the files on disk outside of rsyslog David Lang
> […](#)
> On Tue, 23 Jul 2024, melix-sun wrote: Date: Tue, 23 Jul 2024 23:24:06 -0700 From: melix-sun ***@***.***> Reply-To: rsyslog/rsyslog ***@***.***> To: rsyslog/rsyslog ***@***.***> Cc: Subscribed ***@***.***> Subject: [rsyslog/rsyslog] Duplicated logs uploaded from disk to cloud server (Issue #5419) ### Expected behavior All logs are sync to cloud server in order and no repeated ### Actual behavior Some part of logs are sync repeatedly ### Steps to reproduce the behavior 1. use disk-assisted in-memory queue to sync logs to cloud server 2. logs are written into disk 3. logs in disk are sync to cloud server 4. restart vehicle 5. some part of logs are sync repeatedly ### Environment - rsyslog version:  - platform:  - for configuration questions/issues, include rsyslog.conf and included config files please refer to the attachment for the configuration (duplication details) and debug result (rsyslog-debug), and please kindly help to look at this issue. [rsyslog-debug.zip](https://github.com/user-attachments/files/16358009/rsyslog-debug.zip) [Duplication details.xlsx](https://github.com/user-attachments/files/16358008/Duplication.details.xlsx) if you need more details to figure it out, please kindly let me know, thanks.

Hi David,

Thanks for you reply. the configuration is as below:

$WorkDirectory   /diagUpdate/onlineSyslog # where to place spool files
--
$ActionQueueType   LinkedList   # run asynchronously
$ActionQueueFileName uniqName   # unique name prefix for spool files
$ActionQueueSize 20000
$ActionQueueDiscardMark 16000
$ActionQueueDequeueBatchSize   10000
$ActionQueueHighWaterMark   18000
$ActionQueueLowWaterMark 14000
$ActionQueueMaxFileSize 10m
$ActionQueueMaxDiskSpace   200m   # 200m space limit (use as much   as possible)
$ActionQueueDequeueSlowDown   1000000
$ActionQueueSaveOnShutdown on   # save messages to disk on shutdown
$ActionResumeRetryCount   -1    # infinite retries if host is   down
$ActionQueueTimeoutEnqueue 0
$ModLoad omrelp



The above are all the configurations about sync logs. I am not sure what kind of config do I need to provide.

**Comment by melix-sun** [comment]
Hi ,

Would you please explain how rsyslog delete log once the log has been sync to server from disk queue?

**Comment by davidelang** [comment]
melix-sun wrote:

> Hi ,
>
> Would you please explain how rsyslog delete log once the log has been sync to server from disk queue?

are you talking about the queue file? you can't sync those and expect anything 
sane to happen. they don't get deleted and state is stored in the file (but 
even that gets 'interesting' as getting the state synced to disk doesn't happen 
continually)

David Lang

**Comment by melix-sun** [comment]
> are you talking about the queue file? you can't sync those and expect anything sane to happen. they don't get deleted and state is stored in the file (but even that gets 'interesting' as getting the state synced to disk doesn't happen continually

I mean the logs in disk queue, we set the queue file size as: $ActionQueueMaxFileSize 10m, and we find that as long as the queue file is less than 10M, then the logs in the queue file will be sync repeatedly once restart the system.

**Comment by davidelang** [comment]
melix-sun wrote:

>> are you talking about the queue file? you can't sync those and expect anything sane to happen. they don't get deleted and state is stored in the file (but even that gets 'interesting' as getting the state synced to disk doesn't happen continually
>
> I mean the logs in disk queue, we set the queue file size as: $ActionQueueMaxFileSize 10m, and we find that as long as the queue file is less than 10M, then the logs in the queue file will be sync repeatedly once restart the system.

Yes, if you copy the queue files elsewhere and try to process them, you will 
lose the state information about what messages in the file have already been 
processed. you should not try to do that. Those files are not designed to be 
used the way you are trying to use them.

David Lang

**Comment by melix-sun** [comment]
> melix-sun wrote:
> > are you talking about the queue file? you can't sync those and expect anything sane to happen. they don't get deleted and state is stored in the file (but even that gets 'interesting' as getting the state synced to disk doesn't happen continually I mean the logs in disk queue, we set the queue file size as: $ActionQueueMaxFileSize 10m, and we find that as long as the queue file is less than 10M, then the logs in the queue file will be sync repeatedly once restart the system.
> Yes, if you copy the queue files elsewhere and try to process them, you will lose the state information about what messages in the file have already been processed. you should not try to do that. Those files are not designed to be used the way you are trying to use them. David Lang

I did not copy the file to other place

**Comment by melix-sun** [comment]
Do we need to config the state of the message in disk queue files or it is a default process in rsyslog and no need to config the state for message after sync to server?

**Comment by rgerhards** [comment]
Sorry for the delayed follow-up. rsyslog does not lightly close issues and I am not closing this just because it is old; I reviewed the discussion while working through the backlog.

The thread already has the important queue/RELP semantics answer: disk queue files are internal queue state, the last queue file is not necessarily removed, and rsyslog manages the queue state itself. Seeing data remain in a queue file is not by itself evidence that it will be resent. Duplicates can happen with interrupted delivery semantics, especially around RELP acknowledgment loss, because rsyslog favors at-least-once delivery over silently dropping messages when the sender cannot know whether the receiver processed a batch.

The remaining question required an exact current sequence showing rsyslog resending already-acknowledged queue data without an interrupted shutdown/network/ack scenario, but that evidence was not provided. I am closing this as answered / needs a new reproducible report. If the behavior can still be reproduced on a current rsyslog version, please open a new issue with the exact version, minimal config, shutdown/restart/network sequence, and receiver-side evidence identifying duplicated message IDs.
